File: app/src/main/java/com/elementary/tasks/core/app_widgets/WidgetDataProvider.java
package com.elementary.tasks.core.app_widgets;

import android.app.AlarmManager;
import android.content.Context;

import com.elementary.tasks.birthdays.BirthdayItem;
import com.elementary.tasks.core.utils.Configs;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.TimeCount;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.reminder.models.Reminder;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class WidgetDataProvider {

    public enum WidgetType {
        BIRTHDAY,
        REMINDER
    }

    private List<Item> data;
    private SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
    private int hour, minute;
    private boolean isFeature;
    private Context mContext;

    public WidgetDataProvider(Context context){
        this.mContext = context;
        data = new ArrayList<>();
    }

    public void setTime(int hour, int minute){
        this.hour = hour;
        this.minute = minute;
    }

    public void setFeature(boolean isFeature){
        this.isFeature = isFeature;
    }

    public List<Item> getData(){
        return data;
    }

    public Item getItem(int position){
        return data.get(position);
    }

    public boolean hasReminder(int day, int month, int year){
        boolean res = false;
        for (Item item : data){
            if (res) break;
            int mDay = item.getDay();
            int mMonth = item.getMonth();
            int mYear = item.getYear();
            WidgetType type = item.getType();
            res = mDay == day && mMonth == month && mYear == year && type == WidgetType.REMINDER;
        }
        return res;
    }

    public boolean hasBirthday(int day, int month){
        boolean res = false;
        for (Item item : data){
            int mDay = item.getDay();
            int mMonth = item.getMonth();
            WidgetType type = item.getType();
            if (mDay == day && mMonth == month && type == WidgetType.BIRTHDAY) {
                res = true;
                break;
            }
        }
        return res;
    }

    public void fillArray(){
        data.clear();
        loadBirthdays();
        loadReminders();
    }

    public void loadReminders(){
        List<Reminder> reminderItems = RealmDb.getInstance().getEnabledReminders();
        for (Reminder item : reminderItems) {
            int mType = item.getType();
            long eventTime = item.getDateTime();
            if (!Reminder.isGpsType(item.getType()) && eventTime > 0) {
                Calendar calendar = Calendar.getInstance();
                calendar.setTimeInMillis(eventTime);
                int mDay = calendar.get(Calendar.DAY_OF_MONTH);
                int mMonth = calendar.get(Calendar.MONTH);
                int mYear = calendar.get(Calendar.YEAR);
                if (eventTime > 0) {
                    data.add(new Item(mDay, mMonth, mYear, WidgetType.REMINDER));
                }
                long repeatTime = item.getRepeatInterval();
                long limit = item.getRepeatLimit();
                long count = item.getEventCount();
                boolean isLimited = limit > 0;
                if (isFeature) {
                    Calendar calendar1 = Calendar.getInstance();
                    calendar1.setTimeInMillis(eventTime);
                    if (Reminder.isBase(mType, Reminder.BY_WEEK)) {
                        long days = 0;
                        long max = Configs.MAX_DAYS_COUNT;
                        if (isLimited) max = limit - count;
                        List<Integer> list = item.getWeekdays();
                        do {
                            calendar1.setTimeInMillis(calendar1.getTimeInMillis() + AlarmManager.INTERVAL_DAY);
                            int weekDay = calendar1.get(Calendar.DAY_OF_WEEK);
                            if (list.get(weekDay - 1) == 1) {
                                int sDay = calendar1.get(Calendar.DAY_OF_MONTH);
                                int sMonth = calendar1.get(Calendar.MONTH);
                                int sYear = calendar1.get(Calendar.YEAR);
                                days++;
                                data.add(new Item(sDay, sMonth, sYear, WidgetType.REMINDER));
                            }
                        } while (days < max);
                    } else if (Reminder.isBase(mType, Reminder.BY_MONTH)) {
                        long days = 0;
                        long max = Configs.MAX_DAYS_COUNT;
                        if (isLimited) max = limit - count;
                        do {
                            item.setEventTime(TimeUtil.getGmtFromDateTime(eventTime));
                            eventTime = TimeCount.getInstance(mContext).getNextMonthDayTime(item);
                            calendar1.setTimeInMillis(eventTime);
                            days++;
                            int sDay = calendar1.get(Calendar.DAY_OF_MONTH);
                            int sMonth = calendar1.get(Calendar.MONTH);
                            int sYear = calendar1.get(Calendar.YEAR);
                            data.add(new Item(sDay, sMonth, sYear, WidgetType.REMINDER));
                        } while (days < max);
                    } else {
                        if (repeatTime == 0) continue;
                        long days = 0;
                        long max = Configs.MAX_DAYS_COUNT;
                        if (isLimited) max = limit - count;
                        do {
                            calendar1.setTimeInMillis(calendar1.getTimeInMillis() + repeatTime);
                            mDay = calendar1.get(Calendar.DAY_OF_MONTH);
                            mMonth = calendar1.get(Calendar.MONTH);
                            mYear = calendar1.get(Calendar.YEAR);
                            days++;
                            data.add(new Item(mDay, mMonth, mYear, WidgetType.REMINDER));
                        } while (days < max);
                    }
                }
            }
        }
    }

    public void loadBirthdays(){
        List<BirthdayItem> list = RealmDb.getInstance().getAllBirthdays();
        for (BirthdayItem item : list) {
            Date date = null;
            try {
                date = format.parse(item.getDate());
            } catch (ParseException e) {
                e.printStackTrace();
            }
            if (date != null) {
                Calendar calendar1 = Calendar.getInstance();
                calendar1.setTime(date);
                int bDay = calendar1.get(Calendar.DAY_OF_MONTH);
                int bMonth = calendar1.get(Calendar.MONTH);
                calendar1.setTimeInMillis(System.currentTimeMillis());
                calendar1.set(Calendar.MONTH, bMonth);
                calendar1.set(Calendar.DAY_OF_MONTH, bDay);
                calendar1.set(Calendar.HOUR_OF_DAY, hour);
                calendar1.set(Calendar.MINUTE, minute);
                data.add(new Item(bDay, bMonth, 0, WidgetType.BIRTHDAY));
            }
        }
    }

    public static class Item {
        private int day;
        private int month;
        private int year;
        private WidgetType type;

        public Item(int day, int month, int year, WidgetType type){
            this.day = day;
            this.month = month;
            this.year = year;
            this.type = type;
        }

        public int getYear(){
            return year;
        }

        public void setYear(int year){
            this.year = year;
        }

        public int getMonth(){
            return month;
        }

        public void setMonth(int month){
            this.month = month;
        }

        public int getDay(){
            return day;
        }

        public void setDay(int day){
            this.day = day;
        }

        public WidgetType getType(){
            return type;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/app_widgets/WidgetDataProvider.kt
package com.elementary.tasks.core.app_widgets

import com.elementary.tasks.core.data.models.Reminder
import com.elementary.tasks.core.data.repository.BirthdayRepository
import com.elementary.tasks.core.data.repository.ReminderRepository
import com.elementary.tasks.core.utils.Configs
import com.elementary.tasks.core.utils.datetime.DateTimeManager
import com.elementary.tasks.core.utils.plusMillis
import org.threeten.bp.LocalTime

class WidgetDataProvider(
  private val dateTimeManager: DateTimeManager,
  private val reminderRepository: ReminderRepository,
  private val birthdayRepository: BirthdayRepository
) {

  private val data: MutableList<Item> = ArrayList()
  private var birthdayTime: LocalTime = LocalTime.now()
  private var isFeature: Boolean = false

  enum class WidgetType {
    BIRTHDAY,
    REMINDER
  }

  fun setTime(birthdayTime: LocalTime) {
    this.birthdayTime = birthdayTime
  }

  fun setFeature(isFeature: Boolean) {
    this.isFeature = isFeature
  }

  fun getData(): List<Item> {
    return data
  }

  fun getItem(position: Int): Item {
    return data[position]
  }

  fun hasReminder(day: Int, month: Int, year: Int): Boolean {
    var res = false
    for (item in data) {
      if (res) {
        break
      }
      res = item.day == day && item.month == month && item.year == year &&
        item.type == WidgetType.REMINDER
    }
    return res
  }

  fun hasBirthday(day: Int, month: Int): Boolean {
    var res = false
    for (item in data) {
      if (item.day == day && item.month == month && item.type == WidgetType.BIRTHDAY) {
        res = true
        break
      }
    }
    return res
  }

  fun prepare() {
    data.clear()
    loadBirthdays()
    loadReminders()
  }

  private fun loadReminders() {
    val reminders = reminderRepository.getActiveWithoutGpsTypes()
    for (item in reminders) {
      val mType = item.type
      var eventTime = dateTimeManager.fromGmtToLocal(item.eventTime) ?: continue
      data.add(
        Item(
          eventTime.dayOfMonth,
          eventTime.monthValue,
          eventTime.year,
          WidgetType.REMINDER
        )
      )
      val repeatTime = item.repeatInterval
      val limit = item.repeatLimit.toLong()
      val count = item.eventCount
      val isLimited = limit > 0
      if (isFeature) {
        if (Reminder.isBase(mType, Reminder.BY_WEEK)) {
          var days: Long = 0
          var max = Configs.MAX_DAYS_COUNT
          if (isLimited) {
            max = limit - count
          }
          val list = item.weekdays
          do {
            eventTime = eventTime.plusDays(1)

            val weekDay = dateTimeManager.localDayOfWeekToOld(eventTime.dayOfWeek)
            if (list[weekDay - 1] == 1) {
              days++
              data.add(
                Item(
                  eventTime.dayOfMonth,
                  eventTime.monthValue,
                  eventTime.year,
                  WidgetType.REMINDER
                )
              )
            }
          } while (days < max)
        } else if (Reminder.isBase(mType, Reminder.BY_MONTH)) {
          var days: Long = 0
          var max = Configs.MAX_DAYS_COUNT
          if (isLimited) {
            max = limit - count
          }
          do {
            item.eventTime = dateTimeManager.getGmtFromDateTime(eventTime)
            eventTime = dateTimeManager.getNewNextMonthDayTime(item)
            days++
            data.add(
              Item(
                eventTime.dayOfMonth,
                eventTime.monthValue,
                eventTime.year,
                WidgetType.REMINDER
              )
            )
          } while (days < max)
        } else {
          if (repeatTime == 0L) {
            continue
          }
          var days: Long = 0
          var max = Configs.MAX_DAYS_COUNT
          if (isLimited) {
            max = limit - count
          }
          do {
            eventTime = eventTime.plusMillis(repeatTime)
            days++
            data.add(
              Item(
                eventTime.dayOfMonth,
                eventTime.monthValue,
                eventTime.year,
                WidgetType.REMINDER
              )
            )
          } while (days < max)
        }
      }
    }
  }

  private fun loadBirthdays() {
    val birthdays = birthdayRepository.getAll()
    for (item in birthdays) {
      val date = dateTimeManager.parseBirthdayDate(item.date) ?: continue
      data.add(Item(date.dayOfMonth, date.monthValue, 0, WidgetType.BIRTHDAY))
    }
  }

  class Item(var day: Int, var month: Int, var year: Int, val type: WidgetType)
}
